staging/src/k8s.io/apimachinery/pkg/util/mergepatch/errors.go

```diff
@@ -17,28 +17,31 @@ limitations under the License.
 package mergepatch
 
 import (
-	"errors"
 	"fmt"
+	"net/http"
 	"reflect"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 )
 
 var (
-	ErrBadJSONDoc                           = errors.New("invalid JSON document")
-	ErrNoListOfLists                        = errors.New("lists of lists are not supported")
-	ErrBadPatchFormatForPrimitiveList       = errors.New("invalid patch format of primitive list")
-	ErrBadPatchFormatForRetainKeys          = errors.New("invalid patch format of retainKeys")
-	ErrBadPatchFormatForSetElementOrderList = errors.New("invalid patch format of setElementOrder list")
-	ErrPatchContentNotMatchRetainKeys       = errors.New("patch content doesn't match retainKeys list")
+	ErrBadJSONDoc                           = apierrors.NewBadRequest("invalid JSON document")
```

Comment:
While technically this works, I have slight problems to define the http error code of apiserver http handlers in some apimachinery util package deep in the call stack. But the alternative using some more abstract error types here and to map them to http errors in the handlers does not look appealing either.

Reply:
Good point. I think client code (kubectl?) also calls the functions in this package. It will be confusing if server errors are returned by client-side code. cc @mengqiy

Maybe adding a InterpretPatchError function in the handler to map the abstract error to http errors?

Reply:
with kubectl in mind that's probably better